haystack/nodes/retriever/_openai_encoder.py

```diff
@@ -0,0 +1,144 @@
+import json
+import logging
+import sys
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+import numpy as np
+import requests
+from tqdm.auto import tqdm
+
+from haystack.errors import OpenAIError, OpenAIRateLimitError
+from haystack.nodes.retriever._base_embedding_encoder import _BaseEmbeddingEncoder
+from haystack.schema import Document
+from haystack.utils.reflection import retry_with_exponential_backoff
+from haystack.utils.import_utils import _optional_component_not_installed
+
+if TYPE_CHECKING:
+    from haystack.nodes.retriever import EmbeddingRetriever
+
+USE_TIKTOKEN = False
+if sys.version_info >= (3, 8):
+    USE_TIKTOKEN = True
+
+if USE_TIKTOKEN:
+    try:
+        import tiktoken
+    except (ImportError, ModuleNotFoundError) as ie:
+        _optional_component_not_installed(__name__, "openai", ie)
+else:
+    from transformers import GPT2TokenizerFast, PreTrainedTokenizerFast
```

Comment:
I may add a logger.debug statement here (similar to as what you print below) warning the user that the `tiktoken` library is not available which will change the behavior of the `_OpenAIEmbeddingEncoder` node if the newer generation of embedding models are used.